From e337132af191da52dce407850258bd2e8edb6cd6 Mon Sep 17 00:00:00 2001
From: Shayne Chen <shayne.chen@mediatek.com>
Date: Fri, 12 Apr 2024 11:45:41 +0800
Subject: [PATCH 055/115] mtk: mac80211: defer enabling beacon for MLD AP

Do not enable beacon on the first beacon update (NL80211_CMD_NEW_BEACON)
for MLD AP, let it start from the next beacon update
(NL80211_CMD_SET_BEACON).
This is used to make sure that MLD AP start beacon after all links
finish settings.

Signed-off-by: Shayne Chen <shayne.chen@mediatek.com>
---
 net/mac80211/cfg.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/net/mac80211/cfg.c b/net/mac80211/cfg.c
index 79fd0b51..91002772 100644
--- a/net/mac80211/cfg.c
+++ b/net/mac80211/cfg.c
@@ -1530,7 +1530,7 @@ static int ieee80211_start_ap(struct wiphy *wiphy, struct net_device *dev,
 	}
 
 	link_conf->dtim_period = params->dtim_period;
-	link_conf->enable_beacon = true;
+	link_conf->enable_beacon = !ieee80211_vif_is_mld(&sdata->vif);
 	link_conf->allow_p2p_go_ps = sdata->vif.p2p;
 	link_conf->twt_responder = params->twt_responder;
 	link_conf->he_obss_pd = params->he_obss_pd;
@@ -1612,6 +1612,11 @@ static int ieee80211_start_ap(struct wiphy *wiphy, struct net_device *dev,
 
 	ieee80211_vif_cfg_change_notify(sdata, BSS_CHANGED_SSID);
 	ieee80211_link_info_change_notify(sdata, link, changed);
+	/* for MLD AP, enable_beacon is false during the first beacon set,
+	 * enable it after that. This allows userspace to control the
+	 * beacon enable timing.
+	 */
+	link_conf->enable_beacon = true;
 
 	if (ieee80211_num_beaconing_links(sdata) <= 1)
 		netif_carrier_on(dev);
-- 
2.45.2

